blockchain/consensus/narwhal_cns/src/consensus.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright 2019-2022 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+use async_trait::async_trait;
+use narwhal_fastcrypto::PublicKey;
+use std::fmt::Debug;
+use std::sync::Arc;
+use thiserror::Error;
+
+use forest_blocks::{Block, Tipset};
+use forest_chain::Error as ChainStoreError;
+use forest_chain::Scale;
+use forest_chain::Weight;
+use forest_chain_sync::consensus::Consensus;
+use forest_ipld_blockstore::BlockStore;
+use forest_state_manager::Error as StateManagerError;
+use forest_state_manager::StateManager;
+use fvm_ipld_encoding::Error as ForestEncodingError;
+use fvm_shared::bigint::BigInt;
+use nonempty::NonEmpty;
+
+#[derive(Debug, Error)]
+pub enum NarwhalConsensusError {
+    #[error("Chain store error: {0}")]
+    ChainStore(#[from] ChainStoreError),
+    #[error("StateManager error: {0}")]
+    StateManager(#[from] StateManagerError),
+    #[error("Encoding error: {0}")]
+    ForestEncoding(#[from] ForestEncodingError),
+    #[error("No miner address for validator key: {0}")]
+    NoMinerAddress(PublicKey),
+    #[error("{0}")]
+    Other(String),
+}
+
+/// In Narwhal Consensus we don't gossip blocks, we create deterministic
+/// blocks locally from the certificates we receive from Narwhal, and
+/// append them to our local chain.
+///
+/// The catch is that Narwhal only allows committee members to participate,
```

Comment:
This is a fun challenge. Actually, Mir is facing the same problem right now, it assumes that every subnet participant is a validator (which at this point is fine, but we should think about how to support also full-nodes that do not participate in the consensus, as this is a still requirement for IPC).

Reply:
Narwhal currently has no validation on the contents they broadcast and sign. If it did, by the virtue of everyone (1) checking what they accept from clients and (2) checking the batch contents before they vote on it, then we could at least ensure that everything included is a valid `SignedMessage`. Then we could use the quorum certificate in the Narwhal output to forward these blocks to other nodes on the network, they could check that the signature is over the content they receive. 

However the batches would still not be aligned with Filecoin block limits, so this would work better as a state machine replication protocol that is just an amplification of the core Narwhal committee, than a generator of the Filecoin blockchain.

Maybe the Narwhal quorum certificate could be included in the blocks before their transactions, so everyone knows how many are coming, and then non-validators can check whether the next N transactions arriving in the next M blocks match the signatures of the originating committee.

Alternatively we could extend the Filecoin protocol to gossip votes for each block until a quorum is achieved, like Tendermint, and send them along the blocks to the rest of the network so they can tell which block is legit.

Reply:
On second thought, what I proposed with including the quorum certificates from Narwhal into the blocks would probably not work, because it would still allow the adversary to eliminate a full certificate and its batches from the blocks, and I would not be able to tell anything is missing.